bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthFactoryImpl.java

```diff
@@ -0,0 +1,171 @@
+            @NonNull String clientId, @Nullable String clientSecret, @Nullable String scope,
+            @Nullable Boolean supportsBasicAuth) {
+
+        String uuidHandle = UUID.randomUUID().toString();
+
+        @SuppressWarnings("null")
+        OAuthClientServiceImpl clientImpl = OAuthClientServiceImpl.getInstance(bundleContext, uuidHandle,
+                oAuthStoreHandler, tokenUrl, authorizationUrl, clientId, clientSecret, scope, supportsBasicAuth,
+                tokenExpiresIn_Buffer, httpClientFactory);
+
+        oauthClientServiceCache.put(uuidHandle, clientImpl);
+        return uuidHandle;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public OAuthClientService getOAuthClientService(@NonNull String handle) {
+        OAuthClientServiceImpl clientImpl = null;
+        try {
+            clientImpl = oauthClientServiceCache.get(handle);
+            if (clientImpl == null || clientImpl.isClosed()) {
+                // This happens after reboot, or client was closed without factory knowing; create a new client
+                // the store has the handle/config data
+                clientImpl = OAuthClientServiceImpl.getInstance(bundleContext, handle, oAuthStoreHandler,
+                        tokenExpiresIn_Buffer, httpClientFactory);
+                oauthClientServiceCache.put(handle, clientImpl);
+            }
+        } catch (OAuthException e) {
+            logger.error("Cannot get OAuth client service by handle {}", handle, e);
+        }
+        return clientImpl;
+    }
+
+    @Override
+    public void ungetOAuthService(@NonNull String handle) {
+        OAuthClientServiceImpl clientImpl = oauthClientServiceCache.get(handle);
+        if (clientImpl == null) {
+            logger.error("{} handle not found.  Cannot unregisterOAuthServie", handle);
+            return;
+        }
+        clientImpl.close();
+    }
+
+    @Override
+    public void deleteServiceAndAccessToken(@NonNull String handle) {
+        OAuthClientServiceImpl clientImpl = oauthClientServiceCache.get(handle);
+        if (clientImpl != null) {
+            clientImpl.remove();
+        }
+        oauthClientServiceCache.remove(handle);
+        oAuthStoreHandler.remove(handle);
+    }
+
+    @Override
+    public void deleteAllServicesAndAccessTokens() {
+        oauthClientServiceCache.forEach((handler, clientImpl) -> {
+            if (clientImpl != null) {
+                clientImpl.remove();
+            }
+        });
+        oauthClientServiceCache.clear();
+        oAuthStoreHandler.removeAll();
+    }
+
+    /**
+     * The Store handler is mandatory, but the actual storage service is not.
+     * OAuthStoreHandler will handle when storage service is missing.
+     *
+     * Intended static mandatory 1..1 reference
+     *
+     * @param oAuthStoreHandler
+     */
+    @Reference
+    protected void setoAuthStoreHandler(OAuthStoreHandler oAuthStoreHandler) {
```

Comment:
rename to `setOAuthStoreHandler`

Reply:
Good catch; this was generated by eclipse !